sympy/physics/mechanics/kane.py

```diff
@@ -317,6 +317,17 @@ def _initialize_constraint_matrices(self, config, vel, acc, linear_solver='LU'):
                 self._f_dnh = msubs(acc, udot_zero)
                 self._k_dnh = (acc - self._f_dnh).jacobian(self._udot)
 
+            # The linear velocity constraints must be linear in u, so check'
+            # for u in the components.
+            dy_syms = find_dynamicsymbols(self._f_nh.row_join(self._k_nh))
```

Comment:
If you use [linear_eq_to_matrix](https://docs.sympy.org/latest/modules/solvers/solveset.html#sympy.solvers.solveset.linear_eq_to_matrix) you'll speed up the creation of the matrices, while raising a NonlinearError if the velocity constraints are nonlinear. E.g.
```python
self._k_nh, self._f_nh = linear_eq_to_matrix(vel, self.u)
```

Reply:
Thanks!
But would this not, in logical consequence, entail rewriting _def initialize_constraint_matrices_ completely?

Reply:
Thanks!
But would this not, in logical consequence, entail rewriting _def initialize_constraint_matrices_ completely?

Reply:
> But would this not, in logical consequence, entail rewriting def initialize_constraint_matrices(.....) completely?

It would entail replacing the lines computing the matrix and vector:
```python
self._k_nh = ...jacobian(...)
self._f_nh = ...
```
While the rest of the function can remain unchanged.

The `NonlinearError` would indeed be raised by `linear_eq_to_matrix` with `initialize_constraint_matrices` in the stack trace, i.e. you can also see that it was raised when calling `linear_eq_to_matrix` here. If this error is not clear enough, then an option would be to catch the NonlinearError and update the message. Something like:
```python
try:
    ... = linear_eq_to_matrix(...)
except NonlinearError as e:
    msg = "Some clear message."
    raise NonlinearError(msg) from e
```
However, I would expect that the error message given by `linear_eq_to_matrix` is clear enough.

Reply:
I tried as you suggested:
```
            if self._qdot_u_map is not None:
                vel = msubs(vel, self._qdot_u_map)
            self._f_nh, self._k_nh = linear_eq_to_matrix([vel, self.u])
        #    self._f_nh = msubs(vel, u_zero)
        #    self._k_nh = (vel - self._f_nh).jacobian(self.u)
```
but got this error message:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[9], [line 8](vscode-notebook-cell:?execution_count=9&line=8)
      [6](vscode-notebook-cell:?execution_count=9&line=6) speed_constraint = [i - j for i, j in zip(loesung, u_dep)]
      [7](vscode-notebook-cell:?execution_count=9&line=7) print(len(speed_constraint))
----> [8](vscode-notebook-cell:?execution_count=9&line=8) KM = KanesMethodComplex(N, q_ind=q_ind, u_ind=u_ind, u_dependent=u_dep, kd_eqs=kd, velocity_constraints=speed_constraint)
      [9](vscode-notebook-cell:?execution_count=9&line=9) (fr, frstar) = KM.kanes_equations(BODY, FL)
     [11](vscode-notebook-cell:?execution_count=9&line=11) MM = KM.mass_matrix_full

Cell In[2], [line 235](vscode-notebook-cell:?execution_count=2&line=235)
    [233](vscode-notebook-cell:?execution_count=2&line=233) _validate_coordinates(self.q, self.u)
    [234](vscode-notebook-cell:?execution_count=2&line=234) self._initialize_kindiffeq_matrices(kd_eqs, kd_eqs_solver)
--> [235](vscode-notebook-cell:?execution_count=2&line=235) self._initialize_constraint_matrices(
    [236](vscode-notebook-cell:?execution_count=2&line=236)     configuration_constraints, velocity_constraints,
    [237](vscode-notebook-cell:?execution_count=2&line=237)     acceleration_constraints, constraint_solver)

Cell In[2], [line 305](vscode-notebook-cell:?execution_count=2&line=305)
    [303](vscode-notebook-cell:?execution_count=2&line=303)     if self._qdot_u_map is not None:
    [304](vscode-notebook-cell:?execution_count=2&line=304)         vel = msubs(vel, self._qdot_u_map)
--> [305](vscode-notebook-cell:?execution_count=2&line=305)     self._f_nh, self._k_nh = linear_eq_to_matrix([vel, self.u])
    [306](vscode-notebook-cell:?execution_count=2&line=306) #    self._f_nh = msubs(vel, u_zero)
    [307](vscode-notebook-cell:?execution_count=2&line=307) #    self._k_nh = (vel - self._f_nh).jacobian(self.u)
    [308](vscode-notebook-cell:?execution_count=2&line=308)     # If no acceleration constraints given, calculate them.
    [309](vscode-notebook-cell:?execution_count=2&line=309)     if not acc:

File c:\Users\Peter\anaconda3\envs\sympy-dev\Lib\site-packages\sympy\solvers\solveset.py:2845, in linear_eq_to_matrix(equations, *symbols)
   [2767](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2767) r"""
   [2768](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2768) Converts a given System of Equations into Matrix form.
   [2769](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2769) Here `equations` must be a linear system of equations in
   (...)
   [2842](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2842)     be done *before* calling this routine.
   [2843](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2843) """
   [2844](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2844) if not symbols:
-> [2845](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2845)     raise ValueError(filldedent('''
   [2846](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2846)         Symbols must be given, for which coefficients
   [2847](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2847)         are to be found.
   [2848](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2848)         '''))
   [2850](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2850) # Check if 'symbols' is a set and raise an error if it is
   [2851](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2851) if isinstance(symbols[0], set):

ValueError: 
Symbols must be given, for which coefficients are to be found.
```

Reply:
It doesn't look like you are using the function properly. See the documentation:

https://docs.sympy.org/latest/modules/solvers/solveset.html#sympy.solvers.solveset.linear_eq_to_matrix

Reply:
> It doesn't look like you are using the function properly. See the documentation:
> 
> https://docs.sympy.org/latest/modules/solvers/solveset.html#sympy.solvers.solveset.linear_eq_to_matrix

Now I see, I simply copied Timo, without looking.